packages/agents/src/lib/mcpClient.ts

```diff
@@ -0,0 +1,184 @@
+import { SSEEdgeClientTransport } from "./sseEdge";
+
+import {
+  ToolListChangedNotificationSchema,
+  type ClientCapabilities,
+  type Resource,
+  type Tool,
+  type Prompt,
+  ResourceListChangedNotificationSchema,
+  PromptListChangedNotificationSchema,
+  type ListToolsResult,
+  type ListResourcesResult,
+  type ListPromptsResult,
+  type ServerCapabilities,
+} from "@modelcontextprotocol/sdk/types.js";
+import { Client } from "@modelcontextprotocol/sdk/client/index.js";
+import {
+  SSEClientTransport,
+  type SSEClientTransportOptions,
+} from "@modelcontextprotocol/sdk/client/sse.js";
+
+export class MCPClientConnection {
+  client: Client;
+  transport: SSEClientTransport;
+  connected: boolean;
+  instructions?: string;
+  tools: Tool[];
+  prompts: Prompt[];
+  resources: Resource[];
+  serverCapabilities: ServerCapabilities | undefined;
+
+  constructor(
+    url: URL,
+    private info: ConstructorParameters<typeof Client>[0],
+    opts: {
+      transport: SSEClientTransportOptions;
+      client: ConstructorParameters<typeof Client>[1];
+      capabilities: ClientCapabilities;
+    } = { transport: {}, client: {}, capabilities: {} }
+  ) {
+    this.transport = new SSEEdgeClientTransport(url, opts.transport);
+    this.client = new Client(info, opts.client);
+    this.client.registerCapabilities(opts.capabilities);
+    this.connected = false;
+    this.tools = [];
+    this.prompts = [];
+    this.resources = [];
+  }
+
+  async init() {
+    await this.client.connect(this.transport);
+
+    this.serverCapabilities = await this.client.getServerCapabilities();
+    if (!this.serverCapabilities) {
+      throw new Error(
+        `The MCP Server ${this.info.name} failed to return server capabilities`
+      );
+    }
+
+    const [instructions, tools, resources, prompts] = await Promise.all([
+      this.client.getInstructions(),
+      this.registerTools(),
+      this.registerResources(),
+      this.registerPrompts(),
+    ]);
+
+    this.instructions = instructions;
+    this.tools = tools;
+    this.resources = resources;
+    this.prompts = prompts;
+  }
+
+  /**
+   * Notification handler registration
+   */
+  async registerTools(): Promise<Tool[]> {
+    if (!this.serverCapabilities || !this.serverCapabilities.tools) {
+      return [];
+    }
+
+    if (this.serverCapabilities.tools.listChanged) {
```

Comment:
Oh I see, you have support for list changed - in that case does the client need to tell the server about this in initialization?

Reply:
Yeah, mentioned a bit more here: https://github.com/cloudflare/agents/pull/125#discussion_r2015369764